checkoint coordinator: handle failure on saving zero checkpoint

yumkam opened the pull request:
### Changelog entry <!-- a user-readable short description of changes introduced in this PR -->

...

### Changelog category <!-- remove all except one -->

* Bugfix 
* Not for changelog (changelog entry is not required)

### Additional information

...

yumkam committed:
checkoint coordinator: handle failure to save zero checkpoint

yumkam committed:
more FailedZeroCheckpoint triggers

kardymonds commented on ydb/core/fq/libs/checkpointing/checkpoint_coordinator.cpp:
```diff
@@ -372,11 +372,12 @@ void TCheckpointCoordinator::Handle(const TEvCheckpointCoordinator::TEvScheduleC
     ScheduleNextCheckpoint();
     const auto checkpointsInFly = PendingCheckpoints.size() + PendingCommitCheckpoints.size();
-    if (checkpointsInFly >= Settings.GetMaxInflight() || InitingZeroCheckpoint) {
+    if (checkpointsInFly >= Settings.GetMaxInflight() || (InitingZeroCheckpoint && !FailedZeroCheckpoint)) {
```
Тогда тесты неправильно написаны, есть тест ShouldAbortPreviousCheckpointsIfNodeStateCantBeSaved. 
В нем есть MockRunGraph(), которого видимо не должно быть. А когда он должен приходить непонятно

yumkam replied:
Он должен приходить после сохранения ZeroCheckpoint (а в тесте есть только симуляция одного -- первого -- сохранения чекпоинта -- неуспешная); никакого влияния ни на что это не оказывает, но правильнее его убрать

yumkam replied:
В принципе, ещё можно добавить ассерты в `Handle(EvRunGraph)`.
```
    Y_ABORT_UNLESS(InitingZeroCheckpoint);
    Y_ABORT_UNLESS(!FailedZeroCheckpoint);
```
В норме оно вроде должно работать (`TEvZeroCheckpointDone->SetLoadFromCheckpointMode->Ping(SetLoadFrom...Cookie)->EvRunGraph`), но в тестах что-то налажано и они с этим падают :-|

yumkam replied:
Разобрался с тестами, добавил assert для не-релиза

kardymonds replied:
Так теперь если  ZeroCheckpoint  фейлится, то TEvZeroCheckpointDone не отправляется и  граф не запускается. А чекпойнты начнут идти. Может лучше граф явно зафейлить.

yumkam replied:
Если я правильно понял TODO, то граф работает сразу. Но раньше несохранившийся zero checkpoint ломал чекпоинты вообще, а теперь первый же сохранившийся просто играет роль zero checkpoint.
Если ошибка транзиентная, и просто повторение поможет -- всё отлично.
Если ошибка не транзиентная, и чекпоинты вообще не сохраняются, никакой разницы с тем, что это начало происходить после zero checkpoint: нужны алерты за failure rate.

kardymonds replied:
Как всё запутано) Если граф сразу запускается, тогда InitingZeroCheckpoint означает просто признак необходимости отравить TEvZeroCheckpointDone при первом же успешном чекпойнте. И в условии `checkpointsInFly >= Settings.GetMaxInflight()` все эти флаги не нужны.

yumkam replied:
Там ещё отдельная ветка с восстановлением из checkpoint, там тоже выставляется InitingZeroCheckpoint

yumkam committed:
correct (harmless) ut inaccuracy

... on simulating save failure.
Only one (zero'th) checkpoint save is simulated; as it simulates
failure, EvZeroCheckpointDone is not expected to be sent:
Expected calls sequence is
EvZeroCheckpointDone->EvForwardPing(SetLoadMode...Cookie)->EvRunGraph

yumkam committed:
add (debug for now) asserts and fix ut
